# HG changeset patch
# User Jaroslav Hajek <highegg@gmail.com>
# Date 1274774638 -7200
#      Tue May 25 10:03:58 2010 +0200
# Node ID 9e4407c19e8d16706ad2579e9b45c72983a50865
# Parent  984b8b6bee02bd288b1a19a7601a3676cfa80300
make median more Matlab compatible (bug #29930)

diff --git a/scripts/ChangeLog b/scripts/ChangeLog
--- a/scripts/ChangeLog
+++ b/scripts/ChangeLog
@@ -1,8 +1,13 @@
+2010-05-25  Jaroslav Hajek  <highegg@gmail.com>
+
+	* statistics/base/median.m: Ensure NaNs where needed. Simplify first
+	non-singleton dimension computation.
+
 2010-05-24  Rik <octave@nomad.inbox5.com>
         * general/private/__isequal__.m, image/image.m, 
         statistics/base/ranks.m: Replace calls to deprecated functions.
 
 2010-05-23  Rik <octave@nomad.inbox5.com>
 
         * image/imshow.m: Scale image point color according to colormap. 
         Bug #29926.
diff --git a/scripts/statistics/base/median.m b/scripts/statistics/base/median.m
--- a/scripts/statistics/base/median.m
+++ b/scripts/statistics/base/median.m
@@ -1,11 +1,11 @@
 ## Copyright (C) 1996, 1997, 1998, 1999, 2000, 2004, 2005, 2006, 2007,
 ##               2008, 2009 John W. Eaton
-## Copyright (C) 2009 VZLU Prague
+## Copyright (C) 2009, 2010 VZLU Prague
 ##
 ## This file is part of Octave.
 ##
 ## Octave is free software; you can redistribute it and/or modify it
 ## under the terms of the GNU General Public License as published by
 ## the Free Software Foundation; either version 3 of the License, or (at
 ## your option) any later version.
 ##
@@ -49,27 +49,29 @@
 ## Author: jwe
 
 function retval = median (a, dim)
 
   if (nargin != 1 && nargin != 2)
     print_usage ();
   endif
   if (nargin < 2)
-    dim = [find(size (a) != 1, 1), 1](1); # First non-singleton dim.
+    [~, dim] = max (size (a) != 1); # First non-singleton dim.
   endif
 
   if (numel (a) > 0)
     n = size (a, dim);
     k = floor ((n+1) / 2);
     if (mod (n, 2) == 1)
       retval = nth_element (a, k, dim);
     else
       retval = mean (nth_element (a, k:k+1, dim), dim);
     endif
+    ## Inject NaNs where needed, to be consistent with Matlab.
+    retval(any (isnan (a), dim)) = NaN;
   else
     error ("median: invalid matrix argument");
   endif
 
 endfunction
 
 %!test
 %! x = [1, 2, 3, 4, 5, 6];
